### DISCUSSION TITLE
Online interactive demonstration

### DISCUSSION THREAD

**Discussion by pthom** [epezent/implot#183] [General]
Your library is truly awesome, congrats!

When it come to C++ libraries, most of the time there is most of the time no easy way to get a working demonstration. First, you have to download the code setup the build environment, etc. There, C++ is lagging behind some other languages, especially on the web.
 
However, emscripten can help a lot in this case. In order to tryr to fill this gap, I spent a few hours this afternoon in order to make an online demo of your library.

It is live [here](https://traineq.org/implot_demo/src/implot_demo.html), and the [source code](https://github.com/pthom/implot_demo) for this demo is on github. 

Here are some screenshots:

implot demo:
<img width="718" alt="image" src="https://user-images.githubusercontent.com/7694091/86048896-7a962900-ba51-11ea-843c-a670a30e0577.png">

source code browser with access to the different demos:
<img width="717" alt="image" src="https://user-images.githubusercontent.com/7694091/86048951-8c77cc00-ba51-11ea-837c-fe14390a6871.png">

It runs at 120 fps, except for the benchmark (which I suspect is blocked because it may consume too much memory)

If you are interested, you could may be use this in order to add an interactive demo in your Readme. Feel free to deploy the wasm files on any server of your choice.

Thanks

**Comment by leeoniya** [comment]
nice. though it seems to stress my laptop's thermal management :D

related: #54

**Comment by pthom** [comment]
> nice. though it seems to stress my laptop's thermal management :D
> 
> related: #54

Let me reduce the fps to 60, it will be more gentle!

**Comment by pthom** [comment]
Is it more gentle now? The fps is now chosen by the browser (and varies between 40 to 60 on my laptop)

**Comment by epezent** [comment]
This is too cool! Thank you for putting this together. I've been meaning to add compiled binaries of the demo to the Releases, but I love the idea of a web-based demo. 

> If you are interested, you could may be use this in order to add an interactive demo in your Readme. Feel free to deploy the wasm files on any server of your choice.

I need to familiarize myself emscripten first. Do you mind if I add a link to your demo in our README?

**Comment by pthom** [comment]
> I need to familiarize myself emscripten first. Do you mind if I add a link to your demo in our README?

With pleasure; of course you can. I will leave it on the server. 
Ping me if you want me to rebuild it after some modifications on your side (it's just a matter of  "submodule update" + make + rsync)

**Comment by pthom** [comment]
Hi, I just updated the demo in order because I saw that it could fail on some browsers (safari and edge notably).

It is corrected, and I have tested it succesfuly on firefox (MacOS, Windows, Linux), chrome(MacOS, Window, Linux), safari (MacOS, iOS), as well as on Android.

Also, the code buttons will now always put the corresponding code at the top of the page:
<img width="391" alt="image" src="https://user-images.githubusercontent.com/7694091/86493817-41242d00-bd73-11ea-8565-8cfc4900feec.png">

**Comment by epezent** [comment]
@pthom, I know it's been a while but I finally updated the README to link to your demo. Thanks again for creating and hosting this. Should any major features be added, I will let you know if the demo needs rebuilding. If you ever decide to stop hosting, please let me know. Closing this issue.

**Comment by pthom** [comment]
@epezent : thanks for letting me know! Ping me when you want an update.

**Comment by pthom** [comment]
Just for info, I updated the demo today with the latest code.

**Comment by epezent** [comment]
Thanks @pthom!

The benchmark is still crashing for whatever reason. Were you ever able to investigate? I'm curious if it's something we could fix on our end.

**Comment by bear24rw** [comment]
Does the benchmark still with #88 and compiling with `-DImDrawIdx="unsigned int"`?

**Comment by pthom** [comment]
I have updated the code with the correct compile definition, and the benchmark now works.
(see https://github.com/pthom/implot_demo/commit/64cfb15804f29c9d4d2a43643ed5792d45d43144)

By the way, I also updated to implot current version (which required to Create/Destroy Implot context).

**Comment by epezent** [comment]
Thanks @pthom! I added some new styling stuff to the demo this morning that you may want to update for. Also, if it's not too much trouble could you try merging the ImGui `tables` branch so that the Tables section of the demo compiles. It merges seemlessly with `master` and `docking`.

![image](https://user-images.githubusercontent.com/29577475/91073030-0797d180-e600-11ea-8b41-65c9acafb733.png)

**Comment by pthom** [comment]
@epezent : I updated the implot code and used a mix branch between docking and tables (see this [commit](https://github.com/pthom/implot_demo/commit/0b84659633a9406a31d55bca090093b5c064810f)) 

There were some merge errors in the imgui example files which I deliberately ignored. 

The tables example now works.

**Comment by epezent** [comment]
@pthom , I was wondering if you experience white flickering dots in the ImPlot online demo. I noticed them today. I see them in Chrome and Edge. I checked your other Hello ImGui projects, and didn't see any. I was wondering if somehow ImPlot is causing the issue, but the white dots appear before any ImPlot code has been invoked (i.e. immedialtey when the demo is opened and before any plots are shown). Is this an Emscripten issue?

**Comment by pthom** [comment]
@epezent : I just make a test on several browsers (namely firefox under Windows, linux and osx; chrome under osx and windows; edge under windows; chromium under linux) and I did not experience any flickering on neither of them.

My main machine is a mac, and I am using virtual machines for windows and linux.

Could you share a screenshot or a small video? Are you able to reproduce it on a different PC?

**Comment by epezent** [comment]
Sent you an email. I'll try to reproduce on another machine later.

**Comment by pthom** [comment]
Just a quick message to say that I updated the demo with the latest implot version (+ latest commit on imgui's docking branch), as of 2021/03/10.

**Comment by pthom** [comment]
Hello,

I just updated the demo with the latest implot commit.
Also, I updated the code so that copying code from the editor window works. It was previously not possible, due to security limitations with webAsm. I also added a "copy button", see below:

<img width="585" alt="image" src="https://user-images.githubusercontent.com/7694091/124669582-ec9a1d00-deb2-11eb-8ae9-1efc1ed854fa.png">

**Comment by epezent** [comment]
Once again, thank you @pthom for hosting the demo!

**Comment by pthom** [comment]
Hello,
I just updated the demo with the latest implot version, as of 2022/02/07.

<img width="550" alt="image" src="https://user-images.githubusercontent.com/7694091/152743286-115892b2-5885-4353-b2d3-7b3eb729ca90.png">

**Comment by pthom** [comment]
Hello, 

Another update to the demo today.
Some details:
- The size of the wasm went from 16MB down to 1.8MB
- The demo is now using the `backends` branch
- The code buttons work again (I saw that implot_demo uses a new naming convention `void ShowDemo_DemoName()`, which the demo parses)


![image](https://user-images.githubusercontent.com/7694091/167428391-6fb0f0ae-f9c6-4c70-b473-367897fdb1a2.png)

**Comment by pthom** [comment]
Happy new year!

I just updated the demo today. I now uses imgui V1.89.2, as well as implot (11/25/2022). 
The font rendering should now be much crisper.

[Link](https://traineq.org/implot_demo/src/implot_demo.html)

**Reply by mravaneesh** [reply]
hello @pthom your demo is a lot of help to all of us, but I wanted to know how did you add the checkbox for GPU acceleration in the config section of the ImPlot demo and if you know how to do it then kindly do tell me.

**Reply by pthom** [reply]
@mravaneesh : I did not see a checkbox in the config section. 

Were you talking about this?
<img width="402" alt="image" src="https://user-images.githubusercontent.com/7694091/212190111-d186e30e-fdf7-4625-bd2c-f0a750910551.png">

If so, this is an internal feature of [Hello ImGui](https://github.com/pthom/hello_imgui)

**Comment by pthom** [comment]
Update (Feb 2025): the [interactive demo](https://traineq.org/implot_demo/src/implot_demo.html) now groups the demos for ImPlot and [ImPlot3D](https://github.com/brenocq/implot3d/).

<img width="500" alt="image" src="https://github.com/user-attachments/assets/b2fc06f2-adc9-4c1d-a323-f93eeba2a83f" />

Also, it now provides button to quickly access [implot_demo.cpp](https://github.com/epezent/implot/blob/master/implot_demo.cpp) code on Github. It also provides a button to access the related [implot_demo.py](https://github.com/pthom/imgui_bundle/blob/main/bindings/imgui_bundle/demos_python/demos_implot/implot_demo.py) file (available if using Dear ImGui Bundle python bindings) 

<img width="554" alt="image" src="https://github.com/user-attachments/assets/100a4ddf-a4d4-4c5e-93fa-2762bfcf65bc" />

**Comment by EFeru** [comment]
Hi @pthom ,
Sorry for jumping in. I am really impressed, and I am very curious how you managed to run implot in a browser. I am investigating if I can do similarly and update live data from js into charts that use implot. What do you think?
Thanks for your answer.

**Reply by pthom** [reply]
Thanks for the feedback!

> I am very curious how you managed to run implot in a browser

ImGui and ImPlot can be compiled to webasm / emscripten. This is what I did, and in this case, I used my library [Hello ImGui](https://github.com/pthom/hello_imgui) (on top of ImGui), because it provides a ready to use toolchain for emscripten.

> I am investigating if I can do similarly and update live data from js into charts that use implot. 

It is not that straightforward, but you might investigate

https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html#implement-c-in-javascript

There I found these extracts

> A case where this might be required is if you want to import a large amount of data from JavaScript to be processed by compiled code. For example, the following code allocates a buffer, copies in some data, calls a C function to process the data, and finally frees the buffer.

```
var buf = Module._malloc(myTypedArray.length*myTypedArray.BYTES_PER_ELEMENT);
Module.HEAPU8.set(myTypedArray, buf);
Module.ccall('my_function', 'number', ['number'], [buf]);
Module._free(buf);
```

You may also look at

https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html#calling-javascript-from-c-c
